File: scripts/migrate_fs_labeled.mjs
import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const p = path.join(__dirname, "..", "stdlib", "fs.vibra");
let s = fs.readFileSync(p, "utf8");

function sub(re, repl) {
  s = s.replace(re, repl);
}

sub(
  /(\n      )\$function:\n        args:\n          self: \$self\n        return:/g,
  "$1$function: $self\n      return:",
);
sub(
  /(\n        )\$function:\n          args:\n            self: \$self\n          return:/g,
  "$1$function: $self\n          return:",
);
for (const extra of [
  "start:",
  "segment:",
  "b:",
  "len:",
  "pos:",
  "n:",
  "whence:",
  "mode:",
  "off:",
]) {
  sub(
    new RegExp(
      `(\\n        )\\$function:\\n          args:\\n            self: \\$self\\n            ${extra.replace(":", "\\:")}`,
      "g",
    ),
    `$1$function: $self\n          args:\n            ${extra}`,
  );
}

sub(
  /^(\s{2})\$function:\n        args:\n          p: \$path\n          grant:/gm,
  "$1$function: $path\n    args:\n      grant:",
);
sub(
  /^(\s{2})\$function:\n        args:\n          p: \$path\n          read-grant:/gm,
  "$1$function: $path\n    args:\n      read-grant:",
);
sub(
  /^(\s{2})\$function:\n        args:\n          grant: \$security\.fs-read-grant\n          p:/gm,
  "$1$function: $security.fs-read-grant\n    args:\n      p:",
);
sub(
  /^(\s{2})\$function:\n        args:\n          grant: \$security\.fs-write-grant\n          p:/gm,
  "$1$function: $security.fs-write-grant\n    args:\n      p:",
);

s = s.replaceAll("- $args.p\n", "- $args.subject\n");

fs.writeFileSync(p, s);
console.log("wrote", p);

File: scripts/migrate_fs_labeled.py
"""One-off: migrate stdlib/fs.vibra nested $function to labeled form."""
import pathlib
import re

p = pathlib.Path(__file__).resolve().parent.parent / "stdlib" / "fs.vibra"
text = p.read_text(encoding="utf-8")

def sub(pat, repl, s, flags=0):
    return re.sub(pat, repl, s, flags=flags)

# path.* defs: 6-space base, 8-space $function
s = text
s = sub(
    r"(\n      )\$function:\n        args:\n          self: \$self\n        return:",
    r"\1$function: $self\n      return:",
    s,
)
s = sub(
    r"(\n        )\$function:\n          args:\n            self: \$self\n          return:",
    r"\1$function: $self\n          return:",
    s,
)
# slice / variants with extra args after self
for extra in (
    "start:",
    "segment:",
    "b:",
    "len:",
    "pos:",
    "n:",
    "whence:",
    "mode:",
    "off:",
):
    s = sub(
        rf"(\n        )\$function:\n          args:\n            self: \$self\n            {extra}",
        rf"\1$function: $self\n          args:\n            {extra}",
        s,
    )

# Module-level open-* (2 spaces + $function at col 2)
s = sub(
    r"^(\s{2})\$function:\n        args:\n          p: \$path\n          grant:",
    r"\1$function: $path\n    args:\n      grant:",
    s,
    flags=re.M,
)
s = sub(
    r"^(\s{2})\$function:\n        args:\n          p: \$path\n          read-grant:",
    r"\1$function: $path\n    args:\n      read-grant:",
    s,
    flags=re.M,
)
s = sub(
    r"^(\s{2})\$function:\n        args:\n          grant: \$security\.fs-read-grant\n          p:",
    r"\1$function: $security.fs-read-grant\n    args:\n      p:",
    s,
    flags=re.M,
)
s = sub(
    r"^(\s{2})\$function:\n        args:\n          grant: \$security\.fs-write-grant\n          p:",
    r"\1$function: $security.fs-write-grant\n    args:\n      p:",
    s,
    flags=re.M,
)

# Replace wasm $args.p with $args.subject for path-primary fns (heuristic)
s = s.replace("- $args.p\n", "- $args.subject\n")

p.write_text(s, encoding="utf-8")
print("wrote", p)
